### PR TITLE
Add documentation for the rewriter

### LINKED ISSUE #409: Document missing build steps
In integrating IA2 into zlib I uncovered a number of build steps that aren't currently covered in the docs. We'll need to update the docs to include the following:

* [x] How to run the rewriter, i.e. source changes, compiler args, and rewriter args.
* [x] [`partition-alloc`](https://github.com/immunant/IA2-Phase2/tree/main/runtime/partition-alloc) needs to be included in the final containerized build, and the `--wrap=malloc` linker args need to be added.
* [x] Either `LIBIA2_X86_64` or `LIBIA2_ARM64` defines need to be set for the containerized build.
* [x] Run `pad-tls` on any shared objects that use TLS. We probably also want to document how to determine if a `.so` uses TLS.

### PR DESCRIPTION
**Pull request by randomPoison**
These docs reflect all the steps I needed to get the rewriter working for zlib. Let me know if anything seems wrong, I tried to test everything but I may not have captured all the edge cases for the rewriter.

* Closes #409.

### CODE REVIEW

File: docs/usage.md [hunk 1]
```diff
@@ -7,44 +7,19 @@ they are either assigned one of 15 protection keys or default to the untrusted
...
-(cd $build_directory && $ia2_dir/tools/rewriter/canonicalize_compile_command_paths.py)
-```
-
-## Manual source changes
+### Defining Compartments
 
-### Defining compartments
+The rewriter needs to know which compartment each source file belongs to, and it
+needs each compartment to be initialized in one of the source files. To specify
+the compartment for each file, add a `-DPKEY=$PKEY` argument to the compiler,
```
Comment (kkysen) [inline docs/usage.md:21]:
```suggestion
the compartment for each file, add a `-DPKEY=$PKEY` argument to the compile command,
```

File: docs/usage.md [hunk 2]
```diff
@@ -70,6 +45,76 @@ INIT_RUNTIME(2); // This is the number of pkeys needed
...
+
+The rewriter relies on a `compile_commands.json` file with information about the
+compiler flags used to compile each source file. If your build system does not
+directly support generating a compile commands file, you can use a tool like
+`bear` to record the compiler invocations that your build system makes.
+
+One limitiation is that the rewriter currently does not support having multiple
+compile commands for the same source file. If your build system builds the same
+source file multiple times, e.g. building for both a shared lib and a static
+lib, you'll need to change your build such that it only builds the sources once.
```
Comment (kkysen) [inline docs/usage.md:58]: It "supports" this, in that it'll just warn and choose the first one.

File: docs/usage.md [hunk 3]
```diff
@@ -70,6 +45,76 @@ INIT_RUNTIME(2); // This is the number of pkeys needed
...
+that does this automatically (doing its best to detect paths embedded in args).
+`libclangTooling` also requires compilation databases to be named exactly
+`compile_commands.json`, so this script assumes that as well,
+and must be run in the directory of the `compile_commands.json`, like this:
+
+```sh
+(cd $build_directory && $ia2_dir/tools/rewriter/canonicalize_compile_command_paths.py)
+```
+
+### Running the Rewriter
```
Comment (kkysen) [inline docs/usage.md:74]: Does it make sense to call it this when this is inside of `## Running the Rewriter`?

File: docs/usage.md [hunk 4]
```diff
@@ -70,6 +45,76 @@ INIT_RUNTIME(2); // This is the number of pkeys needed
...
+
+Once you have added the code to initialize each compartment, and have generated
+a `compile_commands.json` with the necessary `PKEY` values specified (via
+`-DPKEY=N`), you can run the rewriter.
+
+Instead of rewriting sources in-place, the rewriter creates a set of new,
+intermediate source (and header) files. Since the rewriter only accepts a list
+of `.c` source files, the set of intermediate headers that will be created is
+controlled by the `--root-directory` and `--output-directory` command-line
+flags. Any file from subdirectories of the root directory which is `#include`d
```
Comment (kkysen) [inline docs/usage.md:84]:
```suggestion
flags. Any file from subdirectories of the root directory that is `#include`d
```

File: docs/usage.md (hunk 3 above)
Comment (randomPoison) [reply docs/usage.md:74]: Oh lol that's a good point. I'll change the first header to "Setup and Rewriter".

File: docs/usage.md (hunk 2 above)
Comment (randomPoison) [reply docs/usage.md:58]: Oh good point. The warnings are pretty easy to miss in the rewriter's output, so I think I still want the warning here, but I'll reword it to be more accurate.

**Review by kkysen — APPROVED** [review]
